### DISCUSSION TITLE
Specifying CSS Color Module Level 4 Colors

### DISCUSSION THREAD

**Discussion by barrycenter** [tailwindlabs/tailwindcss#2218] [Help]
I have a brand color which I would like to specify in tailwindcss which requires a specific color profile to display properly. So, I'd like to provide a color using the CSS Color Module Level 4 color() function. Since this is not widely supported on browsers, it also means I need to specify a fallback color to be used in older browsers as well.

Normally, this would be done in CSS like this:

```
.some-css-class {
    color: #00FF00; /* fallback color */
    color: color(display-p3 0 1 0);
}
```

However, I'm uncertain how to specify a color like this in tailwinds. Ideally, I'd like to set it up in the tailwind.config.js (preferably with 50-900 variants) but I'm non-plussed as to how I should do that.

Is this even possible?

**Comment by barrycenter** [comment]
This is still vexing me as well. Since @RXminuS mentioned this same issue in #4306, I thought I'd mention my findings and my inadequate and hacky workaround.

My first thought was to write a plugin to change the way generateColors() works. If I'm not mistaken, this is the internal function Tailwind uses to generate the `color: rgba(...)` values from the specified hex codes in the color definitions. However, after looking through every example I could find about Tailwind plugins, I never found any that actually modify the base classes output rather than just adding new classes which is not what's needed here. Instead, I need to change the way colors are written for every class with colors (e.g. `bg`, `text`, `border`, `shadow`, `ring`, etc.) and I don't believe I could do this without forking the Tailwind project and changing its internals fundamentally. Which, of course, would make upgrading a serious hassle in the future.

Now for my workaround. I've found that I can specify colors explicitly in my `tailwind.config.js` like this:

```
module.exports = {
  theme: {
    extend: {
      colors: {
        /* CSS fallback colors */
        "bright-green-fallback": {
          "500": "#00FF00",
        },
        /* CSS Color Module Level 4 colors */
        "bright-green-p3": {
          "500": "color(display-p3 0 1 0)",
        }
      }
    }
  }
}
```

Now, you can use the colors like this and have it display properly in all browsers:

`<h1 class="text-bright-green-p3-500 text-bright-green-fallback-500">Bright Green Text</h1>`

However, there are some issues. First, if you leave one of the values out, the text will disappear on Chrome and Firefox if you leave out `bright-green-fallback` and display incorrectly in Safari if you leave out `bright-green-p3`. This is problematic because it means every color needs to be checked in two different browsers to ensure the colors are correct. To avoid potential problems of one being left out, I've created a gigantic, tedious file which looks like this (but with every possible iteration of classes, variants, and dark mode):

```
bg-bright-green-500 { 
  @apply bg-bright-green-p3-500 bg-bright-green-fallback-500 
};
text-bright-green-500 { 
  @apply text-bright-green-p3-500 text-bright-green-fallback-500 
};
hover\:text-bright-green-500 { 
  @apply hover:text-bright-green-p3-500  hover:text-bright-green-fallback-500 
};
dark\:hover\:text-bright-green-500 { 
  @apply dark:hover:text-bright-green-p3-500 dark:hover:text-bright-green-fallback-500 
};
...
```

With this file imported, now I can just write the following and be assured both values are set and that the definitions remain in my `tailwind.config.js` file.

`<h1 class="text-bright-green-500">Bright Green Text</h1>`

As an aside I've done something similar for grays to allow them to automatically adjust for dark mode which is another case where it's easy to leave out a color class if you don't test dark and light modes thoroughly. Like this:

```
bg-dark-auto-300 { @apply bg-gray-300 dark:bg-gray-700 };
...
```

This will be correct about 90% of the time but still allows a designer to apply specific gray and dark:gray classes as desired if it's better for the design.

I have no idea how well either of these solutions will play with the JIT engine that Tailwind is introducing.

Now, back to our CSS Color Module Level 4 colors. The solution I've given above fails miserably when it's used on classes which blend or mix colors like `shadow`, `ring`, or some animations. So you can forget about using them with your fancy new P3 color profile colors. It's not that Safari can't mix P3 colors, it's that mixing them with colors in another color profile is undefined. I believe it would would work great if every color was defined in the same color profile, but that would mean generating a huge file like I did above for every single color and it would be difficult to maintain when new classes are added to Tailwind in the future.

Instead, I have to be cognizant of the repercussions any time I design a component which may use my brand colors (specified in P3). For shadows, I drop back to using only the fallback colors and do a Javascript dance to convert rings to borders if I detect a browser capable of displaying p3 colors. Otherwise, strange things happen, like buttons with a ring border that disappears and never returns if it's specified as a p3 color profile in Safari. (It'll never automatically use the fallback in those instances because the CSS specifications say it should use the more modern color definition so it just disappears entirely when it can't blend it properly.)

### What I would like to see

Ideally, we'd be able to specify more than one definition for each color in our `tailwind.config.js`. Something like this seems ideal:

```
module.exports = {
  theme: {
    extend: {
      colors: {
        /* CSS fallback colors */
        "bright-green": {
          "500": [ "#00FF00", "color(display-p3 0 1 0)" ],
        }
      }
    }
  }
}
```

Though I'd be fine with something like this as well:

```
module.exports = {
  theme: {
    extend: {
      colors: {
        /* CSS fallback colors */
        "bright-green": {
          "500": { 
            hex: "#00FF00",
            color: "color(display-p3 0 1 0)"
          },
        }
      }
    }
  }
}
```

The latter seems to match the current pattern of Tailwind configuration of passing either a string or an object even though it is a bit more verbose.

If multiple definitions could be specified in the configuration, I'd happily create definitions for every single color in `tailwindcss/colors` and leave in the past all of my constant fretting over whether I can or cannot use my brand colors or how to convert classes to keep them from failing in one browser or another.

**Comment by barrycenter** [comment]
For those wondering, "what's the difference?" There are colors that can be specified in other color profiles that cannot be expressed in sRGB which is the default profile used in CSS. To display those colors correctly we need to specify them in a specific color profile. 

Try the following on Safari (which supports the P3 color definitions) to see the difference.

HTML
```
<div class="relative w-64 h-96">
  <div class="absolute p-4 left-0 w-1/2 h-full rgb">RGB</div>
  <div class="absolute p-4 right-0 w-1/2 h-full p3">P3</div>
</div>
```

CSS
```
.rgb {
  background: linear-gradient(
    rgb(255 0 0),
    rgb(255 255 0),
    rgb(0 255 0),
    rgb(0 255 255),
    rgb(0 0 255),
    rgb(255 0 255)
  );
}
.p3 {
  background: linear-gradient(
    color(display-p3 1 0 0),
    color(display-p3 1 1 0),
    color(display-p3 0 1 0),
    color(display-p3 0 1 1),
    color(display-p3 0 0 1),
    color(display-p3 1 0 1)
  );
}
```

Notice how much more vibrant the colors are in the P3 color profile? Those are just some of the colors which cannot be specified with standard CSS. Also, notice the big missing hole where the P3 colors should have been in Chrome and Firefox which don't support CSS Color Module Level 4 Colors yet.

**Comment by RXminuS** [comment]
Yeah, I was thinking about that as a workaround as well. But especially with the new JIT it just feels yucky to have to generate so an enormous blob of colour classes. Ideally, Tailwind would support P3 variants of colours and automatically generate the combination if set. I think this would be hard to do as a plugin from having poked around a bit @adamwathan, would this be something you'd consider adding to the core or should I continue exploring how to do this from the outside?

**Reply by barrycenter** [reply]
The more I look at it, the more impossible it seems to do as a plugin and would likely take significant rewriting of the way Tailwind handles colors in general. 

Color functions can be found scattered throughout Tailwind in the following places:

- `withAlphaVariable.js` _(`withAlphaValue` and `withAlphaVariable` methods add alpha values to rgb and hex colors)_
- `flattenColorPalette.js` _(This is used in many places to create a map of name:value strings)_
- `backgroundColor.js`
- `borderColor.js`
- `divideColor.js`
- `fill.js`
- `gradientColorStops.js`
- `placeholderColor.js`
- `ringColor.js`
- `ringOffsetColor.js`
- `stroke.js`
- `textColor.js`

These all make assumptions about the color being formatted as hex or rgba values and that the colors can be simply converted back and forth between the two. However, we need to be able to specify the color in the proper color profile with a suitable fallback color in the sRGB profile.

Since `flattenColorPalette.js` creates a map of the color value with the color name as the key it is impossible to pass through this function with two different color values for a given color name because they would both have the same key (which is not valid in a map key) so this would need to be refactored to allow multiple values for a color.

Likewise, the `withAlphaValue` and `withAlphaVariable` methods would need to understand the syntax of CSS Color Module Level 4 `color()` function in order to add the appropriate alpha values while adding the ability to handle multiple values per color. (The inability to generate alpha values for colors defined with the `color()` function is what breaks all alpha-reliant classes in my inadequate work-around mentioned earlier.)

Finally, all of the other places would need to be updated (and maybe others since I'm not sure I fully understand the codebase) to output the multiple values correctly.

I don't think this can be done with a plugin and changing those core functions would require updating all other functions which handle colors as well to keep from breaking the regular handling of colors. 

This would need to be done within Tailwind's code.

**Comment by PascalPixel** [comment]
This wouldn't be that hard to implement by overriding the root variables for colors if the browser supports it:

```css
:root {
  --tw-ring-color: rgb(252, 230, 0);
}

@supports (color: color(display-p3 1 1 1 / 1)) {
  :root {
    --tw-ring-color: color(display-p3 0.988 0.902 0);
  }
}
```

To prevent initially having to implement configuration for these, we could initially use a `p3: true;` flag in `tailwind.config.js` and have them auto-generated for p3 only:
```js
let r = 252
let g = 230
let b = 0
let rgb = `rgb(${r}, ${g}, ${b})`
let p3 = `color(display-p3 ${r / 255} ${g / 255} ${b / 255})`
```

Could also have css handle the `calc()`.

P3 does look a lot more vibrant:
![IMG_6679](https://user-images.githubusercontent.com/3094044/168070788-92cf4fd3-91a5-4c1f-8eff-30edcf2a2b3c.jpg)
